### PR TITLE
PLT-6426: fix epoch state endpoint

### PR DESCRIPTION
**Pull request by ana-pantilie**
This PR fixes the new epoch state API meaning that now it should correspond to the [API documentation](https://github.com/input-output-hk/marconi/blob/main/marconi-sidechain/doc/API.adoc#getactivestakepooldelegationbyepoch), see the modified golden test output and the added test.

The changes done to the representation of the `ChainPoint` data in the legacy code (`marconi-sidechain` and `marconi-chain-index-legacy`): the slot and hash need to be partial since `ChainPoint` doesn't always contain this info.

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense and have useful messages
    - [ ] Important changes are reflected in changelog.d of the affected packages
    - [ ] Relevant tickets are mentioned in commit messages
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [ ] Targeting main unless this is a cherry-pick backport
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] If relevant, reference the ADR in the PR and reference the PR in the ADR
    - [ ] Reviewer requested

### CODE REVIEW

File: marconi-chain-index/src/Marconi/ChainIndex/Api/JsonRpc/Endpoint/EpochState.hs [hunk 1]
```diff
@@ -23,9 +30,49 @@ import Network.JsonRpc.Types (JsonRpc, JsonRpcErr)
...
     String
-    (Core.Result EpochState.ActiveSDDByEpochNoQuery)
+    [ActiveSDDResult]
+
+data ActiveSDDResult
+  = ActiveSDDResult
+      !C.PoolId
+      !C.Lovelace
+      !C.SlotNo
+      !(C.Hash C.BlockHeader)
```
Comment (ana-pantilie) [inline marconi-chain-index/src/Marconi/ChainIndex/Api/JsonRpc/Endpoint/EpochState.hs:42]: @berewt : change this to Maybe, and also in legacy, also make the type a record (just here)

File: legacy/marconi-chain-index-legacy/src/Marconi/ChainIndex/Legacy/Indexers/EpochState.hs [hunk 2]
```diff
@@ -477,6 +483,9 @@ instance Buffered EpochStateHandle where
                     isLastImmutableBeforeRollbackable =
                       isBlockRollbackable securityParam (blockNo + 1) (getBlockNoFromChainTip chainTip)
                 when (isRollbackable || isLastImmutableBeforeRollbackable || isFirstEventOfEpoch) $ do
+                  -- The conditions above make it impossible for the ChainPoint to be at genesis,
+                  -- so we should be able to extract the slot and the hash safely.
+                  let (slotNo, blockHeaderHash) = Util.extractSlotNoAndHash chainPoint
```
Comment (ana-pantilie) [inline legacy/marconi-chain-index-legacy/src/Marconi/ChainIndex/Legacy/Indexers/EpochState.hs:488]: I am not 100% sure this reasoning is correct so please let me know if I'm mistaken.

Comment (berewt) [reply legacy/marconi-chain-index-legacy/src/Marconi/ChainIndex/Legacy/Indexers/EpochState.hs:488]: yes, we don't store anything at genesis.

**Review by berewt — DISMISSED** [review]
Just a minor change on the JSON operators.

File: legacy/marconi-chain-index-legacy/test-lib/Gen/Marconi/ChainIndex/Legacy/Types.hs [hunk 3]
```diff
@@ -71,6 +73,13 @@ nonEmptySubset s = do
 genSlotNo :: (Hedgehog.MonadGen m) => m C.SlotNo
 genSlotNo = C.SlotNo <$> Gen.word64 (Range.linear 10 1000)
 
+genMaybeSlotNo :: (Hedgehog.MonadGen m) => m (Maybe C.SlotNo)
+genMaybeSlotNo =
+  Gen.frequency
+    [ (95, Just <$> genSlotNo)
+    , (5, pure Nothing)
+    ]
```
[the remark this answers is not in the document]
Comment (ana-pantilie) [reply legacy/marconi-chain-index-legacy/test-lib/Gen/Marconi/ChainIndex/Legacy/Types.hs:81]: Hmm, good catch.

File: marconi-sidechain/src/Marconi/Sidechain/Api/Routes.hs [hunk 4]
```diff
@@ -405,38 +417,51 @@ instance FromJSON BurnTokenEventResult where
...
-            <*> v .: "lovelace"
-            <*> (C.SlotNo <$> v .: "slotNo")
-            <*> v .: "blockHeaderHash"
+            <$> v
+              .: "poolId"
+            <*> v
+              .: "lovelace"
+            <*> (fmap C.SlotNo <$> v .:? "slotNo")
+            <*> v
+              .: "blockHeaderHash"
```
[the remark this answers is not in the document]
Comment (berewt) [reply marconi-sidechain/src/Marconi/Sidechain/Api/Routes.hs:464]: Shouldn't the blockHeaderHash as well be optional (`.:?`)?

File: marconi-sidechain/src/Marconi/Sidechain/Api/Routes.hs [hunk 5]
```diff
@@ -445,45 +470,47 @@ instance ToJSON ActiveSDDResult where
...
 instance FromJSON NonceResult where
   parseJSON =
     let parseResult v = do
           NonceResult
             <$> (Ledger.Nonce <$> v .: "nonce")
-            <*> (C.SlotNo <$> v .: "slotNo")
-            <*> v .: "blockHeaderHash"
+            <*> (fmap C.SlotNo <$> v .:? "slotNo")
+            <*> v
+              .: "blockHeaderHash"
```
[the remark this answers is not in the document]
Comment (berewt) [reply marconi-sidechain/src/Marconi/Sidechain/Api/Routes.hs:505]: Same question

**Review by berewt — APPROVED** [review]
Unless I get it wrong, the block header hash should still be set as optional, this aside, it's ready to ship.

File: marconi-sidechain/src/Marconi/Sidechain/Api/Routes.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ana-pantilie) [reply marconi-sidechain/src/Marconi/Sidechain/Api/Routes.hs:464]: Oops, I missed these when I added my changes! Sorry, thank you for noticing.

**Pull request merged**